File: src/main/java/org/radarcns/stream/AbstractStreamWorker.java
package org.radarcns.stream;

import java.lang.Thread.UncaughtExceptionHandler;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import org.apache.kafka.streams.KafkaStreams;
import org.apache.kafka.streams.errors.StreamsException;
import org.radarbase.topic.KafkaTopic;
import org.radarcns.config.ConfigRadar;
import org.radarcns.config.KafkaProperty;
import org.radarcns.config.RadarPropertyHandler;
import org.radarcns.config.SingleStreamConfig;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public abstract class AbstractStreamWorker implements StreamWorker, UncaughtExceptionHandler {
    private static final Logger logger = LoggerFactory.getLogger(AbstractStreamWorker.class);

    public static final String OUTPUT_LABEL = "_output";
    public static final Duration TIME_WINDOW_COMMIT_INTERVAL_DEFAULT = Duration.ofSeconds(30);

    private final List<StreamDefinition> streamDefinitions;
    protected SingleStreamConfig config;
    protected ConfigRadar allConfig;
    protected int numThreads;
    protected KafkaProperty kafkaProperty;
    protected StreamMaster master;
    protected List<KafkaStreams> streams;

    public AbstractStreamWorker() {
        streamDefinitions = new ArrayList<>();
    }

    /**
     * Create a stream from input to output topic. By using this method, and {@link
     * #getStreamDefinitions()} automatically gets updated.
     *
     * @param input input topic name
     * @param output output topic name
     */
    protected void defineStream(String input, String output) {
        defineStream(input, output, null);
    }

    /**
     * Create a stream from input to output topic. By using this method, and {@link
     * #getStreamDefinitions()} automatically gets updated.
     *
     * @param input input topic name
     * @param output output topic name
     * @param window time windows size, null if none.
     */
    protected void defineStream(String input, String output, Duration window) {
        streamDefinitions.add(
                new StreamDefinition(new KafkaTopic(input), new KafkaTopic(output), window));
    }

    /**
     * Create a sensor stream from input topic to a "[input]_output" topic. By using this method,
     * and {@link #getStreamDefinitions()} automatically gets updated.
     *
     * @param input input topic name
     */
    protected void defineSensorStream(String input) {
        defineStream(input, input + OUTPUT_LABEL, null);
    }

    /**
     * Create a set of sensor streams, for each of the RADAR standard time frames. An input topic
     * {@code my_input} will create, e.g., {@code my_input_10sec}, {@code my_input_10min} output
     * topics.
     *
     * @param input topic to stream from
     */
    protected void defineWindowedSensorStream(String input) {
        defineWindowedSensorStream(input, input);
    }

    /**
     * Create a set of sensor streams, for each of the RADAR standard time frames. An input topic
     * {@code my_input} with output base {@code my_output} will create, e.g., {@code
     * my_output_10sec}, {@code my_output_10min} output topics.
     *
     * @param input topic to stream from
     * @param outputBase base topic name to stream to
     */
    protected void defineWindowedSensorStream(String input, String outputBase) {
        streamDefinitions.addAll(allConfig.getStream().getTimeWindows(config).stream()
                .map(w -> new StreamDefinition(
                        new KafkaTopic(input),
                        new KafkaTopic(w.getTopicLabel(outputBase)),
                        Duration.ofMillis(w.getIntervalInMilliSec()),
                        allConfig.getStream().getCommitIntervalForTimeWindow(w)
                ))
                .collect(Collectors.toList()));
    }

    @Override
    public Stream<StreamDefinition> getStreamDefinitions() {
        return streamDefinitions.stream();
    }

    @Override
    public void configure(StreamMaster streamMaster, RadarPropertyHandler properties,
            SingleStreamConfig singleConfig) {
        this.kafkaProperty = properties.getKafkaProperties();
        this.allConfig = properties.getRadarProperties();
        this.numThreads = allConfig.getStream().threadsByPriority(singleConfig.getPriority());
        this.config = singleConfig;
        this.master = streamMaster;
        this.initialize();
    }

    /**
     * Starts the stream and notify the StreamMaster.
     */
    @Override
    public void start() {
        if (streams != null) {
            throw new IllegalStateException("Streams already started. Cannot start them again.");
        }
        streams = createStreams();

        if (streams == null) {
            throw new IllegalStateException("Streams are not initialized during doStart");
        }
        streams.forEach(stream -> {
            stream.setUncaughtExceptionHandler(this);
            stream.start();
        });

        master.notifyStartedStream(this);
    }

    protected abstract List<KafkaStreams> createStreams();

    /**
     * Close the stream and notify the StreamMaster.
     */
    @Override
    public void shutdown() {
        logger.info("Shutting down {} stream", getClass().getSimpleName());

        closeStreams();

        master.notifyClosedStream(this);
    }

    protected void closeStreams() {
        if (streams != null) {
            streams.forEach(KafkaStreams::close);
            streams = null;
        }
        doCleanup();
    }

    protected abstract void doCleanup();

    protected abstract void initialize();

    /**
     * Handles exceptions that have been uncaught. It is called when a StreamThread is
     * terminating due to an exception.
     */
    @Override
    public void uncaughtException(Thread t, Throwable e) {
        logger.error("Thread {} has been terminated due to {}", t.getName(), e.getMessage(), e);

        closeStreams();

        if (e instanceof StreamsException) {
            master.restartStream(this);
        } else {
            master.notifyCrashedStream(getClass().getSimpleName());
        }
    }
}

File: src/main/kotlin/org/radarbase/stream/AbstractStreamWorker.kt
package org.radarbase.stream

import org.apache.kafka.streams.KafkaStreams
import org.apache.kafka.streams.errors.StreamsException
import org.apache.kafka.streams.errors.StreamsUncaughtExceptionHandler
import org.radarbase.config.KafkaProperty
import org.radarbase.config.RadarBackendConfig
import org.radarbase.config.RadarConfigHandler
import org.radarbase.config.SingleStreamConfig
import org.radarbase.topic.KafkaTopic
import org.slf4j.LoggerFactory
import java.time.Duration
import java.util.concurrent.CopyOnWriteArrayList
import java.util.stream.Stream

/**
 * Abstract implementation of [StreamWorker] that provides common logic for managing Kafka Streams.
 * This class handles lifecycle management (start, shutdown), error handling, and defines the structure
 * for configuring one or more streams.
 */
abstract class AbstractStreamWorker : StreamWorker, StreamsUncaughtExceptionHandler {
    internal val streamDefinitions = CopyOnWriteArrayList<StreamDefinition>()
    protected lateinit var config: SingleStreamConfig
    protected lateinit var allConfig: RadarBackendConfig
    protected var numThreads: Int = 0
    protected lateinit var kafkaProperty: KafkaProperty
    protected lateinit var master: StreamMaster
    protected var streams: List<KafkaStreams>? = null

    protected fun defineStream(input: String, output: String) {
        defineStream(input, output, null)
    }

    protected fun defineStream(input: String, output: String, window: Duration?) {
        streamDefinitions.add(
            StreamDefinition(KafkaTopic(input), KafkaTopic(output), window),
        )
    }

    protected fun defineStream(def: StreamDefinition) {
        streamDefinitions.add(def)
    }

    protected fun defineSensorStream(input: String) {
        defineStream(input, input + OUTPUT_LABEL, null)
    }

    protected fun defineWindowedSensorStream(input: String) {
        defineWindowedSensorStream(input, input)
    }

    protected fun defineWindowedSensorStream(input: String, outputBase: String) {
        // Since TimeWindowMetadata is in org.radarbase.stream package, we use it directly.
        // If it's not visible, we may need to use a fully qualified name or a helper.
        TimeWindowMetadata.entries.forEach { w ->
            streamDefinitions.add(
                StreamDefinition(
                    KafkaTopic(input),
                    KafkaTopic(w.getTopicLabel(outputBase)),
                    Duration.ofMillis(w.intervalInMilliSec),
                    allConfig.stream!!.getCommitIntervalForTimeWindow(w),
                ),
            )
        }
    }

    override fun getStreamDefinitions(): Stream<StreamDefinition> {
        return streamDefinitions.stream()
    }

    override fun configure(
        streamMaster: StreamMaster,
        properties: RadarConfigHandler,
        singleConfig: SingleStreamConfig,
    ) {
        this.kafkaProperty = properties.kafkaProperties
        this.allConfig = properties.radarProperties
        this.numThreads = allConfig.stream!!.threadsByPriority(singleConfig.priority)
        this.config = singleConfig
        this.master = streamMaster
        this.initialize()
    }

    override fun start() {
        assert(streams == null) { "Streams already started. Cannot start them again." }

        streams = createStreams() ?: throw IllegalStateException("Streams are not initialized during start")

        streams?.forEach { stream ->
            stream.setUncaughtExceptionHandler(this as StreamsUncaughtExceptionHandler?)
            stream.start()
        }

        master.notifyStartedStream(this)
    }

    protected abstract fun createStreams(): List<KafkaStreams>?

    override fun shutdown() {
        logger.info("Shutting down {} stream", javaClass.simpleName)
        closeStreams()
        master.notifyClosedStream(this)
    }

    protected fun closeStreams() {
        streams?.forEach { it.close() }
        streams = null
        doCleanup()
    }

    protected abstract fun doCleanup()

    protected abstract fun initialize()

    override fun handle(e: Throwable?): StreamsUncaughtExceptionHandler.StreamThreadExceptionResponse? {
        logger.error("Thread has been terminated due to {} - {}", e?.message, e)
        closeStreams()
        if (e is StreamsException) {
            master.restartStream(this)
            // TODO check whether this is the right way to handle this
            return StreamsUncaughtExceptionHandler.StreamThreadExceptionResponse.REPLACE_THREAD
        } else {
            master.notifyCrashedStream(javaClass.simpleName)
            // TODO check whether this is the right way to handle this
            return StreamsUncaughtExceptionHandler.StreamThreadExceptionResponse.SHUTDOWN_APPLICATION
        }
    }

    companion object {
        private val logger = LoggerFactory.getLogger(AbstractStreamWorker::class.java)
        const val OUTPUT_LABEL = "_output"

        @JvmField
        val TIME_WINDOW_COMMIT_INTERVAL_DEFAULT: Duration = Duration.ofSeconds(30)
    }
}
